File: fixtures/python/type_inference/discriminated_union/discriminated_union.py
"""Discriminated union narrowing via tuple isinstance and kind field."""

from typing import Union


class Circle:
    kind = "circle"

    def __init__(self, radius: float) -> None:
        self.radius = radius


class Rectangle:
    kind = "rectangle"

    def __init__(self, width: float, height: float) -> None:
        self.width = width
        self.height = height


def area(shape: Union[Circle, Rectangle]) -> float:
    if isinstance(shape, Circle):
        return 3.14 * shape.radius * shape.radius
    return shape.width * shape.height


def describe(shape: Union[Circle, Rectangle]) -> str:
    if isinstance(shape, (Circle, Rectangle)):
        if shape.kind == "circle":
            return "round"
        return "angular"
    return "unknown"


def main() -> None:
    print(area(Circle(1.0)))
    print(describe(Rectangle(2.0, 3.0)))

File: fixtures/java/type_inference/discriminated_union/discriminated_union.java
public sealed interface Shape permits Circle, Rectangle {

    record Circle(double radius) implements Shape {}

    record Rectangle(double width, double height) implements Shape {}

    static double area(Shape shape) {
        if (shape instanceof Circle c) {
            return 3.14 * c.radius() * c.radius();
        } else if (shape instanceof Rectangle r) {
            return r.width() * r.height();
        }
        return 0.0;
    }

    static String describe(Shape shape) {
        if (shape instanceof Circle) {
            return "round";
        }
        return "angular";
    }
}
